src/main/scala/firrtl/passes/NewLowerTypes.scala

```diff
@@ -0,0 +1,466 @@
+// See LICENSE for license details.
+
+package firrtl.passes
+
+import firrtl.analyses.InstanceKeyGraph
+import firrtl.annotations.{CircuitTarget, MemoryInitAnnotation, MemoryRandomInitAnnotation, ModuleTarget, ReferenceTarget}
+import firrtl.{CircuitForm, CircuitState, DependencyAPIMigration, InstanceKind, Kind, MemKind, PortKind, RenameMap, SymbolTable, Transform, UnknownForm, Utils}
+import firrtl.ir._
+import firrtl.options.Dependency
+import firrtl.stage.TransformManager.TransformDependency
+
+import scala.collection.mutable
+
+/** Flattens Bundles and Vecs.
+  * - Some implicit bundle types remain, but with a limited depth:
+  *   - the type of a memory is still a bundle with depth 2 (mem -> port -> field), see [[MemPortUtils.memType]]
+  *   - the type of a module instance is still a bundle with depth 1 (instance -> port)
+  */
+object NewLowerTypes extends Transform with DependencyAPIMigration {
+  override def prerequisites: Seq[TransformDependency] = Seq(
+    Dependency(RemoveAccesses), // we require all SubAccess nodes to have been removed
+    Dependency(CheckTypes),     // we require all types to be correct
+    Dependency(InferTypes),     // we require instance types to be resolved (i.e., DefInstance.tpe != UnknownType)
+    Dependency(ExpandConnects)  // we require all PartialConnect nodes to have been expanded
+  )
+  override def optionalPrerequisiteOf: Seq[TransformDependency]  = Seq.empty
+  override def invalidates(a: Transform): Boolean = a match {
+    case ResolveFlows => true // we generate UnknownFlow for now (could be fixed)
+    case _ => false
+  }
+
+  /** Delimiter used in lowering names */
+  val delim = "_"
+  /** Expands a chain of referential [[firrtl.ir.Expression]]s into the equivalent lowered name
+    * @param e [[firrtl.ir.Expression]] made up of _only_ [[firrtl.WRef]], [[firrtl.WSubField]], and [[firrtl.WSubIndex]]
+    * @return Lowered name of e
+    * @note Please make sure that there will be no name collisions when you use this outside of the context of LowerTypes!
+    */
+  def loweredName(e: Expression): String = e match {
+    case e: Reference => e.name
+    case e: SubField => s"${loweredName(e.expr)}$delim${e.name}"
+    case e: SubIndex => s"${loweredName(e.expr)}$delim${e.value}"
+  }
```

Comment:
I copied this function from `LowerTypes` because it is used by other passes, so I cannot change the argument type, since I want to try and replace `LowerTypes` without a deprecation cycle. However I think it is a good idea to add a better error message than a match error.